tests/test_math_interpolator.hh

```diff
@@ -0,0 +1,158 @@
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNU Emacs; see the file COPYING. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#include "tests.hh"
+#include "test_math.hh"
+#include "math/interpolator.hh"
+#include "representations/calculator_spherical_expansion.hh"
+
+#ifndef TESTS_TEST_MATH_INTERPOLATOR_HH_
+#define TESTS_TEST_MATH_INTERPOLATOR_HH_
+
+namespace rascal {
+  using Vector_t = math::Vector_t;
+  using Matrix_t = math::Matrix_t;
+  using Vector_Ref = math::Vector_Ref;
+  using Matrix_Ref = math::Matrix_Ref;
+
+  using AbsoluteErrorMethod = math::ErrorMethod<math::ErrorMetric_t::Absolute>;
+  using RelativeErrorMethod = math::ErrorMethod<math::ErrorMetric_t::Relative>;
+
+  static double convert_to_scalar(const double x) { return x; }
+  static double convert_to_scalar(const math::Matrix_t & mat) {
+    return mat(0, 0);
+  }
+
+  // These functions calculate the error between the interpolated function and
+  // the interpolation on a set of reference points for different
+
+  template <class ErrorMethod, class ScalarInterpolator>
+  static double compute_intp_error(std::shared_ptr<ScalarInterpolator> & intp,
+                                   std::function<double(double)> ref_func,
+                                   const math::Vector_Ref & ref_points) {
+    math::Vector_t intp_vals = math::Vector_t::Zero(ref_points.size());
+    math::Vector_t intp_refs = math::Vector_t::Zero(ref_points.size());
+    for (int i{0}; i < ref_points.size() - 1; i++) {
+      intp_vals(i) = convert_to_scalar(intp->interpolate(ref_points(i)));
+      intp_refs(i) = ref_func(ref_points(i));
+    }
+    return ErrorMethod::compute_global_error(Vector_Ref(intp_vals),
+                                             Vector_Ref(intp_refs));
+  }
+
+  template <class ErrorMethod, class ScalarInterpolator>
+  static double
+  compute_intp_derivative_error(std::shared_ptr<ScalarInterpolator> & intp,
+                                std::function<double(double)> derivative_func,
+                                const math::Vector_Ref & ref_points) {
+    math::Vector_t intp_vals = math::Vector_t::Zero(ref_points.size());
+    math::Vector_t intp_refs = math::Vector_t::Zero(ref_points.size());
+    for (int i{0}; i < ref_points.size() - 1; i++) {
+      intp_vals(i) =
+          convert_to_scalar(intp->interpolate_derivative(ref_points(i)));
+      intp_refs(i) = derivative_func(ref_points(i));
+    }
+    return ErrorMethod::compute_global_error(Vector_Ref(intp_vals),
```

Comment:
Is this cast needed? Should this not be on the receiving end if it is accessed as a reference or not?

Reply:
Well ok then :)